src/Picker/PickerBuilder.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2017 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\Picker;
+
+use Knp\Menu\FactoryInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Picker factory.
+ *
+ * @author Andreas Schempp <https://github.com/aschempp>
+ */
+class PickerBuilder implements PickerBuilderInterface
+{
+    /**
+     * @var FactoryInterface
+     */
+    private $menuFactory;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    /**
+     * @var RequestStack
+     */
+    private $requestStack;
+
+    /**
+     * @var PickerProviderInterface[]
+     */
+    private $providers = [];
+
+    /**
+     * Constructor.
+     *
+     * @param FactoryInterface $menuFactory
+     * @param RouterInterface  $router
+     * @param RequestStack     $requestStack
+     */
+    public function __construct(FactoryInterface $menuFactory, RouterInterface $router, RequestStack $requestStack)
+    {
+        $this->menuFactory = $menuFactory;
+        $this->router = $router;
+        $this->requestStack = $requestStack;
+    }
+
+    /**
+     * Adds picker providers to the factory.
+     *
+     * @param PickerProviderInterface $provider
+     */
+    public function addProvider(PickerProviderInterface $provider)
+    {
+        $this->providers[] = $provider;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function create(PickerConfig $config)
+    {
+        $providers = array_filter(
+            $this->providers,
+            function (PickerProviderInterface $provider) use ($config) {
+                return $provider->supportsContext($config->getContext());
+            }
+        );
+
+        if (empty($providers)) {
+            return null;
+        }
+
+        return new Picker(
+            $this->menuFactory,
+            $providers,
+            $config
+        );
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function createFromJson($json)
```

Comment:
I would actually add `createFromUrlValue()` and `saveForUrlValue()` or something like this. You can keep the others if you like but this would allow us using `gzencode()` etc. in these methods (or anything else in the future) to transparently and backwards compatible improve the size of the content :)